### ISSUE TITLE
High memory leak using hwdec=auto on intel devices when libvdpau-va-gl1 is installed

### ISSUE THREAD

**Issue by mgcrea** [mpv-player/mpv#3931]
### mpv version and platform

Broken on:

```
mpv git-51fd8f6 (C) 2000-2016 mpv/MPlayer/mplayer2 projects
 built on Fri Dec  9 03:51:07 UTC 2016
```

No issues on:

```
mpv git-343f5ca (C) 2000-2016 mpv/MPlayer/mplayer2 projects
 built on Mon Sep 12 15:19:57 UTC 2016
```

Also no issues for any version when using `--hwdec=no`

Up-to-date ubuntu LTS:

```
Distributor ID:	Ubuntu
Description:	Ubuntu 16.04.1 LTS
Release:	16.04
Codename:	xenial
```

```
# DISPLAY=:0 glxinfo | grep version
server glx version string: 1.4
client glx version string: 1.4
GLX version: 1.4
    Max core profile version: 3.3
    Max compat profile version: 3.0
    Max GLES1 profile version: 1.1
    Max GLES[23] profile version: 3.1
OpenGL core profile version string: 3.3 (Core Profile) Mesa 11.2.0
OpenGL core profile shading language version string: 3.30
OpenGL version string: 3.0 Mesa 11.2.0
OpenGL shading language version string: 1.30
OpenGL ES profile version string: OpenGL ES 3.1 Mesa 11.2.0
OpenGL ES profile shading language version string: OpenGL ES GLSL ES 3.10
```

```
# DISPLAY=:0 vainfo
error: XDG_RUNTIME_DIR not set in the environment.
libva info: VA-API version 0.39.4
libva info: va_getDriverName() returns 0
libva info: Trying to open /usr/lib/x86_64-linux-gnu/dri/i965_drv_video.so
libva info: Found init function __vaDriverInit_0_39
libva info: va_openDriver() returns 0
vainfo: VA-API version: 0.39 (libva 1.7.3)
vainfo: Driver version: Intel i965 driver for Intel(R) Skylake - 1.7.3
vainfo: Supported profile and entrypoints
      VAProfileMPEG2Simple            :	VAEntrypointVLD
      VAProfileMPEG2Simple            :	VAEntrypointEncSlice
      VAProfileMPEG2Main              :	VAEntrypointVLD
      VAProfileMPEG2Main              :	VAEntrypointEncSlice
      VAProfileH264ConstrainedBaseline:	VAEntrypointVLD
      VAProfileH264ConstrainedBaseline:	VAEntrypointEncSlice
      VAProfileH264ConstrainedBaseline:	VAEntrypointEncSliceLP
      VAProfileH264Main               :	VAEntrypointVLD
      VAProfileH264Main               :	VAEntrypointEncSlice
      VAProfileH264Main               :	VAEntrypointEncSliceLP
      VAProfileH264High               :	VAEntrypointVLD
      VAProfileH264High               :	VAEntrypointEncSlice
      VAProfileH264High               :	VAEntrypointEncSliceLP
      VAProfileH264MultiviewHigh      :	VAEntrypointVLD
      VAProfileH264MultiviewHigh      :	VAEntrypointEncSlice
      VAProfileH264StereoHigh         :	VAEntrypointVLD
      VAProfileH264StereoHigh         :	VAEntrypointEncSlice
      VAProfileVC1Simple              :	VAEntrypointVLD
      VAProfileVC1Main                :	VAEntrypointVLD
      VAProfileVC1Advanced            :	VAEntrypointVLD
      VAProfileNone                   :	VAEntrypointVideoProc
      VAProfileJPEGBaseline           :	VAEntrypointVLD
      VAProfileJPEGBaseline           :	VAEntrypointEncPicture
      VAProfileVP8Version0_3          :	VAEntrypointVLD
      VAProfileVP8Version0_3          :	VAEntrypointEncSlice
      VAProfileHEVCMain               :	VAEntrypointVLD
      VAProfileHEVCMain               :	VAEntrypointEncSlice
```

Might related to a bump of the bundled `libva` in the ppa `mpv-tests`.

### Reproduction steps

Start an instance and play video one after the other (eg. via IPC) with hardware decoding enabled

```
/usr/bin/mpv --quiet --idle --keep-open --no-osc --no-border --no-window-dragging --hwdec=auto --no-stop-screensaver --fullscreen --fs-screen=all --ontop --ao=alsa --audio-device=alsa/hdmi --volume=80 --input-ipc-server=/tmp/mpv_socket_fullscreen --msg-level=ipc=v
```

Might be related to https://github.com/mpv-player/mpv/issues/3313

### Expected behavior

Does not eat all RAM

### Actual behavior

Eat all RAM (~5MB for every file), in shared memory.
After ~6H, with a video played every 30s, I get an OOM crash of a few services (memory+swap full).

available goes down while shared goes up.

```
              total        used        free      shared  buff/cache   available
Mem:           3841        1416         125        1015        2299        1136
Swap:          3984          83        3901
```

### Log file

http://sprunge.us/EEIF

### Sample files

Any MP4 video with proper hardware acceleration

**Comment by mc4man** [comment]
Don't see that at all using recent libva/vaapi. How many of those 700+ vid plays were unique?
Here looping 6 vids of varing resolutions/complexity  about 30 sec each for 40 min. there's no sign of a memory leak as seen thru pidstat creating a report every 20 sec 120 times.
Basically mpv no vid was
 VSZ       RSS   
976656   53536

1st 6 reports after playback starts
VSZ           RSS 
1416732   80616
1417244   82572
1417244   83748
1417248   82148
1417248   82240 
1426976   85148

Last 6
 VSZ          RSS  
1428576   88460
1428576   87744
1428576   87800
1428576   88396
1428064   86444
1428064   87524

Noting I did not go exposing the Osc which does seem to increase mem to some extent per use.

**Comment by mgcrea** [comment]
Hey @mc4man, we've been reproducing the issue very easily on our side on multiple devices (a dozen), looping on at most 10 videos.

But the tricky thing is that the memory hog does not seam to be on the mpv side, but on the Xorg process. Monitoring shared memory show it very quickly `watch -n1 free -m`. Also it looks like killing `mpv` won't properly release it, only a `service lightdm restart` does.

We think it's a mesa bug impacting 16.04 (we did not reproduce on an up-to-date 14.04.5 system).

What is bad though is that we did not had the issue before doing a simple `apt update; apt upgrade`, and now our devices are all up-to-date but broken. Not sure what is the right move.

The padoka PPA would probably fix it but it looks quite unstable (and we were thrilled to switch to 16.04 to finally drop experimental oibafs PPA...).

**Comment by mc4man** [comment]
I can't see any increase here overall after running the loop for 40 min,  plus after closing mpv virtually no change.
Before - Mem:           7895         353        4123         192        3419        7007
After -   Mem:           7895         365        4119         183        3410        7004

For not mpv's issue you could try reverting libva, vaapi to 16.04 defaults or maybe go forwards
padoka has a 0.13 stable mesa - 
https://launchpad.net/~paulo-miguel-dias/+archive/ubuntu/pkppa
Or just to test yourself  i've this I had created for personal use several weeks ago.
https://launchpad.net/~mc3man/+archive/ubuntu/tests3

Basically the same mesa, former from git, later from 17.04, mine not a good longterm choice as xserver 0.19 is going in there next month..

**Comment by CounterPillow** [comment]
Interesting that Ubuntu ships mesa 11.2.0, considering that branch already had 2 bugfix releases (namely 11.2.1 and 11.2.2), which seem to include va changes and at least one fix for a (possibly unrelated) memleak.

**Comment by mgcrea** [comment]
Indeed I'm a bit frustrated by the "instability" of an up-to-date LTS. Do you know how often mesa updates are supposed to be backported?

We've tested the padoka stable PPA and it also leaks.
However there is no leaks with the padoka dev PPA. So at least that's a good thing.

We're currently trying to reproduce the issue on a clean base system and to provide you a simple test-case to reproduce it.

**Comment by mgcrea** [comment]
Looks like we can't reproduce it anymore with `mpv git-17d6ba7`. Will close for now, and reopen if I can provide a reproducible test case.

**Comment by mgcrea** [comment]
I've successfully reproduced the issue today on several Nuc6i3 devices.

It looks like the culprit is the installation of `libvdpau-va-gl1` package.

Working log before installation of the package:

http://sprunge.us/YUQD

Broken log after installation of the package:

http://sprunge.us/jCcE

The only difference I spot is this line:

```
[   0.796][v][vdpau] Error when calling vdp_device_create_x11: 1
```

So it looks like the leak might comes from this function (trying to use vdpau with vaapi backend).

We've setup a small JS tool to easily reproduce the leak, just watch for fast increasing shared memory:

https://gitlab.com/carlipa/node-video-ipc

**Comment by lachs0r** [comment]
Don’t use libvdpau-va-gl.

**Comment by mgcrea** [comment]
That's what we will do.

Just not that the [ArchLinux wiki](https://wiki.archlinux.org/index.php/Hardware_video_acceleration#Installation) does recommend to setup both backends:

```
Tip: It is recommended to install and configure both VA-API and VDPAU to achieve support in different scenarios, e.g. Flash doesn't support VA-API but can use the VDPAU VA-API backend.
```

So many users following theses guidelines will end up with broken system on a clean Ubuntu LTS.

**Comment by mc4man** [comment]
@ mgcrea
The only good reason at this point  to have libvdpau-va-gl installed is for vaapi in mplayer, period.
As far as mpv it shouldn't  come into play unless you either have nvidia drivers but no vdpau-hwaccel (libavcodec) support or if using Intel gpu with a vo=vdpau. 
Obviously you'd want to avoid both of those scenarios.
(- mpv that's statically linked to ffmpeg will not pass the default waf configure check statement for vdpau-hwaccel, in at least some distros,  but that can be worked around.

**Comment by wm4** [comment]
I can't find actually anything wrong in those logs. The libvdpau-va-gl1 driver is only a problem if mpv picks up vdpau (for decoding or display),  and as long as you use ``--hwdec=auto`` or just ``--hwdec`` (and obviously don't force ``--vo=vdpau``), mpv shouldn't pick up that driver. So I'm still not sure what's going on there.

Btw. for nvidia cards the trends seems to go towards ``--hwdec=cuda`` being the best choice.

**Comment by mgcrea** [comment]
Just a small update to let you know that removing the package `libvdpau-va-gl` was not a proper fix.

I am now encountering memory leaks on intel hardware with the following message:

```
Failed to open VDPAU backend libvdpau_va_gl.so: cannot open shared object file: No such file or directory
```

For info, looking for vdpau packages:
```
# dpkg -l | grep vdpau
ii  libvdpau1:amd64                             1.1.1-3ubuntu1                                amd64        Video Decode and Presentation API for Unix (libraries)
ii  mesa-vdpau-drivers:amd64                    1:13.0.3-3~x~padoka0                          amd64        Mesa VDPAU video acceleration drivers
ii  vdpau-driver-all:amd64                      1.1.1-3ubuntu1                                amd64        Video Decode and Presentation API for Unix (driver metapackage)
ii  vdpau-va-driver:amd64                       0.7.4-5                                       amd64        VDPAU-based backend for VA API
```

```
# aptitude why libvdpau1
i   mpv Depends libvdpau1 (>= 0.2)
# aptitude why mesa-vdpau-drivers
i   mpv Recommends mesa-vdpau-drivers | nvidia-vdpau-driver | nvidia-driver-binary | nvidia-current | vdpau-driver
# aptitude why vdpau-va-driver
i   mpv             Recommends i965-va-driver | libva-intel-vaapi-driver | va-driver
i A vdpau-va-driver Provides   va-driver   
```

Not sure what is wrong. But there is something very strange going on here.

**Comment by mgcrea** [comment]
Updating again, it looks like this new leak does not come from mpv. Though it still tries to access vdpau, it does not seem to be producing a leak. So false alarm, sorry for the trouble. Will look more into it.

**Comment by lachs0r** [comment]
FWIW I know that the intel xorg DDX causes memory leakage and system freezes with OpenGL applications on my laptop (HD4600). Using modesetting fixes this and many other problems.

**Comment by mgcrea** [comment]
After more investigation, it seems to be confirmed that there is still another memory leak produced by mpv without the package `libvdpau-va-gl`, it's not a proper fix. Switching from `--hwdec=auto` to `--hwdec=vaapi` did fix the leak.

It seems that there is indeed something wrong in the hwdec fallback memory management. Trying vdpau seems to produce a leak.

**Comment by mc4man** [comment]
If you want try this build, don't add ppa, just dl the .deb & install with dpkg. Note that it will only be there a few days. (has vdpau-hwaccel disabled
https://launchpad.net/~mc3man/+archive/ubuntu/quick1/+packages

**Comment by mgcrea** [comment]
@mc4man thanks for the build, should I try with the `libvdpau-va-gl` package (that does easily produce a big leak)?

**Comment by mc4man** [comment]
No,  that package is of almost no value & very few would even have it
installed
